File: templates/r/consort.R
## CONSORT 队列入组流程图（ggplot2 手搓盒+箭头）—— 卡片 11_医学流行病学常用图/CONSORT流程图_Consort.md
## 改 # >>> PARAM 的人数即可。可独立运行： Rscript templates/r/consort.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_consort"; TITLE <- "Study flow (CONSORT-style)"
BOX <- pal("cat_main")[2]     # 盒填充
assessed <- 1200; excluded <- 350; enrolled <- assessed - excluded
arm_a <- 430; arm_b <- 420; ana_a <- 410; ana_b <- 405
excl_txt <- paste0("Excluded (n=", excluded, ")\n• Not meeting criteria (n=210)",
                   "\n• Declined (n=90)\n• Other (n=50)")
# <<< PARAM ------------------------------------------------------

boxes <- data.frame(
  x = c(3.2, 7.3, 3.2, 1.8, 4.6, 1.8, 4.6),
  y = c(11, 9.4, 8.4, 6.0, 6.0, 4.3, 4.3),
  w = c(4.4, 4.2, 4.4, 3.0, 3.0, 3.0, 3.0),
  h = c(1.0, 1.6, 0.9, 0.9, 0.9, 0.9, 0.9),
  fc = c(BOX, "#F0F0F0", BOX, BOX, BOX, BOX, BOX),
  lab = c(sprintf("Assessed for eligibility\n(n=%d)", assessed), excl_txt,
          sprintf("Enrolled / Randomized\n(n=%d)", enrolled),
          sprintf("Arm A\nAllocated (n=%d)", arm_a), sprintf("Arm B\nAllocated (n=%d)", arm_b),
          sprintf("Analyzed (n=%d)", ana_a), sprintf("Analyzed (n=%d)", ana_b)))
seg <- data.frame(x = c(3.2, 3.2, 1.8, 4.6, 1.8, 4.6),
                  xe = c(3.2, 5.2, 1.8, 4.6, 1.8, 4.6),
                  y = c(10.5, 9.7, 7.2, 7.2, 5.55, 5.55),
                  ye = c(8.9, 9.7, 6.45, 6.45, 4.75, 4.75))
p <- ggplot() +
  geom_tile(data = boxes, aes(x, y, width = w, height = h), fill = boxes$fc,
            color = "black", linewidth = .4) +
  geom_text(data = boxes, aes(x, y, label = lab), size = 2.5, lineheight = .9) +
  geom_segment(data = seg, aes(x = x, xend = xe, y = y, yend = ye),
               arrow = arrow(length = unit(1.6, "mm")), linewidth = .4) +
  geom_segment(aes(x = 1.8, xend = 4.6, y = 7.2, yend = 7.2), linewidth = .4) +
  coord_cartesian(xlim = c(0, 10), ylim = c(3.5, 12)) +
  labs(title = TITLE) +
  theme_void() + theme(plot.title = element_text(face = "bold", hjust = .5, size = 10))
save_gg(p, NAME, w = 140, h = 150)

File: templates/python/consort.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""CONSORT 队列入组流程图 —— 卡片 11_医学流行病学常用图/CONSORT流程图_Consort.md
matplotlib 手搓盒+箭头；改 # >>> PARAM 的人数即可。可独立运行。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import matplotlib.pyplot as plt
from matplotlib.patches import FancyBboxPatch, FancyArrowPatch
from _common import set_theme, save_fig, pal
set_theme()

# >>> PARAM ------------------------------------------------------
NAME = "tpl_consort"; TITLE = "Study flow (CONSORT-style)"
BOX = pal("cat_main")[1]      # 盒填充色（浅蓝）
# 各阶段人数与文案
assessed = 1200; excluded = 350
excl_reasons = ["Not meeting criteria (n=210)", "Declined (n=90)", "Other (n=50)"]
enrolled = assessed - excluded
arm_a, arm_b = 430, 420
ana_a, ana_b = 410, 405
FIGSIZE = (5.4, 5.6)
# <<< PARAM ------------------------------------------------------

fig, ax = plt.subplots(figsize=FIGSIZE); ax.set_xlim(0, 10); ax.set_ylim(0, 12)
ax.axis("off"); ax.set_title(TITLE, fontweight="bold")

def box(x, y, w, h, text, fc=BOX):
    ax.add_patch(FancyBboxPatch((x - w / 2, y - h / 2), w, h, fc=fc, ec="black",
                                lw=.8, boxstyle="round,pad=0.02,rounding_size=0.08"))
    ax.text(x, y, text, ha="center", va="center", fontsize=7.5)

def arrow(x1, y1, x2, y2):
    ax.add_patch(FancyArrowPatch((x1, y1), (x2, y2), arrowstyle="-|>",
                                 mutation_scale=10, lw=.9, color="black"))

box(3.2, 11, 4.4, 1.0, f"Assessed for eligibility\n(n={assessed})")
box(7.3, 9.4, 4.2, 1.4, "Excluded (n={})\n".format(excluded) + "\n".join("• " + r for r in excl_reasons),
    fc="#F0F0F0")
arrow(3.2, 10.5, 3.2, 8.9); arrow(3.2, 9.7, 5.2, 9.7)
box(3.2, 8.4, 4.4, 0.9, f"Enrolled / Randomized\n(n={enrolled})")
arrow(3.2, 7.95, 3.2, 7.2); arrow(1.8, 7.2, 1.8, 6.5); arrow(4.6, 7.2, 4.6, 6.5)
ax.plot([1.8, 4.6], [7.2, 7.2], color="black", lw=.9)
box(1.8, 6.0, 3.0, 0.9, f"Arm A\nAllocated (n={arm_a})")
box(4.6, 6.0, 3.0, 0.9, f"Arm B\nAllocated (n={arm_b})")
arrow(1.8, 5.55, 1.8, 4.8); arrow(4.6, 5.55, 4.6, 4.8)
box(1.8, 4.3, 3.0, 0.9, f"Analyzed (n={ana_a})")
box(4.6, 4.3, 3.0, 0.9, f"Analyzed (n={ana_b})")
save_fig(fig, NAME)
